### PR TITLE
Updated stencil classes for kaluga 0.4.0

### PR DESCRIPTION
**Pull request by ChristoferAlexander**
#20 

Updates Date and Color to KalugaDate and KalugaColor

### CODE REVIEW

**Comment by ChristoferAlexander** [comment]
@avdyushin I have some issues with the KalugaDate being an interface now. The Observable and UninitializedObservable generated classes along with their mapper classes have an input of type:
`typealias KotlinObject = AnyObject & Equatable`

There are a couple of ways to tackle this. 

- We could check in the Observable / UninitializedObservable if the input is instance of Equatable and do not do the comparisons in case it is not. From the code I can not fully understand the implications of that so I will need your help there. 

- Make the KalugaDate abstract class so we can still extend for testing but will generate an object for iOS that is Equatable

- Remove KalugaDate observable support as I see very few use case that we want to observe a Date in UI level (usually we observe a formatted String version)

@thoutbeckers tagging you in since you proposed the solutions

**Comment by thoutbeckers** [comment]
> We could check in the Observable / UninitializedObservable if the input is instance of Equatable and do not do the comparisons in case it is not

@ChristoferAlexander Observables already do equality checks before emiting, so they should not be needed. So the question still is, where does the requirement of `Equatable` come from? (in other words, where in the code)

**Comment by ChristoferAlexander** [comment]
> > We could check in the Observable / UninitializedObservable if the input is instance of Equatable and do not do the comparisons in case it is not
> 
> @ChristoferAlexander Observables already do equality checks before emiting, so they should not be needed. So the question still is, where does the requirement of `Equatable` come from? (in other words, where in the code)

Following the whole chain:

```
struct SomeView: View {
    ...
    @ObservedObject private var stringText: StringObservable
    ...
    init(...) {
        let viewModel = SomeViewModel(...)
        viewModel = LifecycleViewModel(viewModel)
        stringText = StringObservable(accountViewModel.nameLabelText)
        }
}

Where:
StringObservable = typealias StringObservable = MappedWithDefaultObservable<NSString, String, StringMapper>

[MappedWithDefaultObservable](https://github.com/splendo/kaluga-swiftui/blob/0cb11c7a663929dc307731399aa50c8b55dd4766/stencils/Observable.stencil)
```

As far as I understand some kind of mapping happens in the init of views from FlowInitializedObservable to the native [ObservableObject](https://developer.apple.com/documentation/combine/observableobject)

**Comment by thoutbeckers** [comment]
This is not an answer to why `Equatable` is needed. An easy way would be to remove `Equatable` from the definition and point to the code that breaks.

File: stencils/Subject.stencil [hunk 1]
```diff
@@ -8,7 +8,7 @@ typealias BoolSubject = MappedSubject<KotlinBoolean, Bool, BoolMapper>
 typealias IntSubject = MappedSubject<KotlinInt, Int, IntMapper>
 typealias FloatSubject = MappedSubject<KotlinFloat, Float, FloatMapper>
 typealias DoubleSubject = MappedSubject<KotlinDouble, Double, DoubleMapper>
-typealias DateSubject = MappedSubject<KalugaDate, Foundation.Date, DateMapper>
+// typealias DateSubject = MappedSubject<KalugaDate, Foundation.Date, DateMapper>
```
Comment (avdyushin) [inline stencils/Subject.stencil:11]: Why it's commented? I will break for sure existing code which uses this modules.
Maybe you have to use DefaultKalugaDate here?

**Comment by avdyushin** [comment]
> This is not an answer to why `Equatable` is needed. An easy way would be to remove `Equatable` from the definition and point to the code that breaks.

In first place `Equatable` was needed to ignore duplicate updates inside observable (by simple ignoring if new object the same as last received)

**Comment by avdyushin** [comment]
But this code was written long time ago, when Kaluga reported duplicated values:

https://github.com/splendo/kaluga-swiftui/blob/770204870ec82af924ac9036e02cbab53e677c1f/stencils/Observable.stencil#L56-L58

Maybe we can retest and if that's not a case (duplicates filtered inside kaluga) we can remove that

File: stencils/Subject.stencil (hunk 1 above)
Comment (ChristoferAlexander) [reply stencils/Subject.stencil:11]: The commended out code is so the generated code could compile for an xBike release but will not be the final product. I did not try DefaultKalugaDate as it made no sense to me.

**Comment by ChristoferAlexander** [comment]
@avdyushin I updated the templates to not use equatable. Could you have a look?

File: stencils/UninitializedSubject.stencil [hunk 2]
```diff
@@ -7,9 +7,9 @@ typealias BoolUninitializedSubject = MappedUninitializedSubject<KotlinBoolean, B
 typealias IntUninitializedSubject = MappedUninitializedSubject<KotlinInt, Int, IntOptionalMapper>
 typealias FloatUninitializedSubject = MappedUninitializedSubject<KotlinFloat, Float, FloatOptionalMapper>
 typealias DoubleUninitializedSubject = MappedUninitializedSubject<KotlinDouble, Double, DoubleOptionalMapper>
-typealias DateUninitializedSubject = MappedUninitializedSubject<KalugaDate, Foundation.Date, DateOptionalMapper>
+// typealias DateUninitializedSubject = MappedUninitializedSubject<KalugaDate, Foundation.Date, DateOptionalMapper>
```
Comment (thoutbeckers) [inline stencils/UninitializedSubject.stencil:10]: can this be removed now?

**Review by avdyushin — APPROVED** [review]

**Comment by avdyushin** [comment]
Before merging this we have to create tag with current master like: kaluga-0.3.0

**Comment by ChristoferAlexander** [comment]
> Before merging this we have to create tag with current master like: kaluga-0.3.0

Sorry @avdyushin, I completely forgot to merge this one. Not exactly sure how to tag it, could you tag it for me please?

**Comment by thoutbeckers** [comment]
@ChristoferAlexander https://github.com/splendo/kaluga-swiftui/tags or make a release there

**Comment by ChristoferAlexander** [comment]
@thoutbeckers, @avdyushin https://github.com/splendo/kaluga-swiftui/releases

**Comment by Daeda88** [comment]
This actually doesnt compile right now. Making some changes to fix that

**Comment by ChristoferAlexander** [comment]
> 

Thanks @Daeda88, noticed yesterday also when I was about to merge but was quite late and left it for today. I initially fixed the generated code locally and transferred the fixes to. the templates. Probably I missed those two.

**Comment by ChristoferAlexander** [comment]
Tested on xBike

**Pull request merged**